### PR TITLE
CLI: Stats command - fix incorrect CPU % reporting

### PR DESCRIPTION
**Pull request by dkbennett**
## Summary of the Pull Request
The CPU calculation is incorrect due to precpu_stats not being populated from the docker request. This is due to the combination of "stream=false" and "one-shot=true" in the request parameters. In this combination, docker engine does not populate precpu_stats. The fix is to remove "one-shot=true" from the request, which causes docker engine to do two samples internally and then have valid population for precpu_stats.

Since docker does an internal sampling of ~1s, and docker does not support batch container requests (one container id per request), running stats on all active containers now has a performance issue of 1s * N containers. This fix also runs the stats requests for all containers in parallel to mitigate this performance issue, and adds a generous E2E test to verify the parallel processing and performance is as-expected and does not regress.

This fix also corrects some minor differences in the stats calculation in the CLI end for edge cases we will probably never hit but aligning it to be exact for the sake of being as correct as possible with the Docker CLI implementation.

## PR Checklist

- [ ] **Closes:** Link to issue #xxx
- [x] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [x] **Tests:** Added/updated if needed and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated if needed
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/wsl/) and link it here: #xxx

## Detailed Description of the Pull Request / Additional comments

Three changes were made here:
1) The stats call in the docker request dropped "one-shot=true". This, when combined with "stream=false" caused docker to not populate precpu_stats. They were zeroed out (confirmed in debugger). Removing this line caused docker engine to do an internal 1s block to capture the prior sample. With the one-shot parameter removed, the precpu_stats were now populated (also confirmed in debugger).

2) The stats calculation had two adjustments to align it exactly with Docker's calculation:
* The "onlineCpus" was changed to have a fallback to percpu_usage size, as docker does. This is because in older docker APIs the stats can report onlineCpus as 0. In this case we had different behavior so this was fixed. The debugger confirmed that this is not actually the case with the APIs we are using so this is only a defensive change in case the API behavior were to change in the future, we should not break.
* The system delta & cpudelta check was using >= while docker uses >. This also should not have any meaningful effect, but for exact correctness with Docker's algorithm it is being fixed.

3) Stats are requested per-container and are now in parallel to account for the ~1 sampling time per container. With all requests being run simultaneously the actual runtime of the stats command should be within a few seconds, even with many containers. A test was added to verify this that is intentionally generous to account for a potentially stressed test machine.

The first change is the real fix to the CPU issue. The second is defensive alignment for edge cases that we should not in practice hit, but might someday encounter, and the third is a mitigation for the potential scaling issue the first change introduced.

## Validation Steps Performed
New parallel stats command with 20 'sleep infinity' containers to verify the sampling time for a large number of containers is relatively close to ~1s. This timing is very generous given the test machine may be stressed. Currently it allows for up to 13s for it to run and be considered still running in parallel.  If this test starts having random failures we can revisit the parameters or consider removing the test.

1 Ran stress tests with 1 cpu and 16 cpus:
* `wslc run --rm alpine sh -c "apk add -q stress-ng && stress-ng --cpu 16 --timeout 60"`
* `wslc run --rm alpine sh -c "apk add -q stress-ng && stress-ng --cpu 1 --timeout 60"`
2 Ran stats in another window to confirm values

Confirmed CPU% was around 100% for the --cpu 1 case and 1600% for the --cpu 16 case. This is correct with Docker's CPU% reporting as 100% per core. 1 fully loaded cpu should be 100%, and 16 should be around 1600%. Excess over these expectations is also expected and includes the process's own overhead and init process consuming additional CPU.

image

Also here is a manual run of 20 containers and the stats command clearly running far less than 20s the sequential command would have.
image

### CODE REVIEW

File: src/windows/wslcsession/DockerHTTPClient.cpp [hunk 1]
```diff
@@ -408,7 +408,10 @@ docker_schema::ContainerStats DockerHTTPClient::ContainerStats(const std::string
 {
     auto url = URL::Create("/containers/{}/stats", Id);
     url.SetParameter("stream", false);
-    url.SetParameter("one-shot", true);
+
+    // Intentionally omit one-shot=true: the Docker engine blocks internally for ~1s
+    // to collect a prior sample, and returns a single response with both cpu_stats and
+    // precpu_stats correctly populated — giving a valid delta for CPU % calculation.
```
[the remark this answers is not in the document]
Comment (dkbennett) [reply src/windows/wslcsession/DockerHTTPClient.cpp:414]: lol no this was the bug

[the remark this answers is not in the document]
Comment (dkbennett) [reply src/windows/wslcsession/DockerHTTPClient.cpp:414]: However, this is a valid scalability concern since we do this request for each container.  In order to get the diff on perf we do need two samples so I will look into fixing the perf concern separately. Concurrent fetching of stats seems like the best approach rather than sequential.

Edit: I am adding concurrency to the CLI command now to address this, along with test. It will bloat the PR significantly but I believe it is correct to address the scalability issue with the PR that introduces it if possible.

[the remark this answers is not in the document]
Comment (dkbennett) [reply src/windows/wslcsession/DockerHTTPClient.cpp:414]: This is fixed with the latest commit, adding parallelization to the stats command and introducing a test to verify that functionality is correct.

File: src/windows/wslc/tasks/ContainerTasks.cpp [hunk 2]
```diff
@@ -495,41 +495,84 @@ void ShowContainerStats(CLIExecutionContext& context)
...
+        // Formula matches Docker CLI: https://github.com/docker/cli/blob/master/cli/command/container/stats_helpers.go
         double cpuPercent = 0.0;
         const auto cpuDelta = static_cast<double>(stats.cpu_stats.cpu_usage.total_usage) -
                               static_cast<double>(stats.precpu_stats.cpu_usage.total_usage);
         const auto systemDelta =
             static_cast<double>(stats.cpu_stats.system_cpu_usage) - static_cast<double>(stats.precpu_stats.system_cpu_usage);
-        const auto onlineCpus = stats.cpu_stats.online_cpus > 0 ? stats.cpu_stats.online_cpus : 1u;
-        if (systemDelta > 0.0 && cpuDelta >= 0.0)
+
+        // When online_cpus is 0 (older API responses), fall back to percpu_usage array length — matches Docker CLI behavior.
```
Comment (OneBlue) [inline src/windows/wslc/tasks/ContainerTasks.cpp]: Is that something that can happen with our version of docker ? If not, I'd just recommend removing that block entirely since the API response we get should be consistent

File: test/windows/wslc/e2e/WSLCE2EContainerStatsTests.cpp [hunk 3]
```diff
@@ -109,6 +109,71 @@ class WSLCE2EContainerStatsTests
...
+            VERIFY_IS_TRUE(result.StdoutContainsSubstring(name), std::format(L"Container '{}' not found in stats output", name).c_str());
+        }
+
+        WEX::Logging::Log::Comment(
+            std::format(L"Stats for {} containers completed in {}ms (parallel max: {}ms, sequential min: {}ms)", c_containerCount, elapsedMs, c_parallelMaxMs, c_sequentialMinMs)
+                .c_str());
+
+        VERIFY_IS_GREATER_THAN_OR_EQUAL(
+            elapsedMs, c_minExpectedMs, L"Stats completed suspiciously fast - Docker sampling may not be occurring");
+        VERIFY_IS_LESS_THAN(
```
Comment (OneBlue) [inline test/windows/wslc/e2e/WSLCE2EContainerStatsTests.cpp]: Sadly this is one of the cases that's difficult to test in a reliable way. I don't recommend running this timing check in the CI, since it's likely to break if Defender / Windows Update triggers right as this test runs.

File: src/windows/wslc/tasks/ContainerTasks.cpp [hunk 4]
```diff
@@ -495,41 +495,84 @@ void ShowContainerStats(CLIExecutionContext& context)
         }
     }
 
-    // Build stats as a json array first for later filtering or display either as json or table format.
-    nlohmann::json statsJson = nlohmann::json::array();
+    // Fetch stats for all containers concurrently. The Docker engine blocks for ~1s per
```
Comment (OneBlue) [inline src/windows/wslc/tasks/ContainerTasks.cpp]: Hmm that's too bad that the request hangs for a second. Maybe we could implement a workaround in the service to issue N requests in parallel and return it to the client, but the current API doesn't support that so I think this is OK for now

File: test/windows/wslc/e2e/WSLCE2EContainerStatsTests.cpp (hunk 3 above)
Comment (dkbennett) [reply test/windows/wslc/e2e/WSLCE2EContainerStatsTests.cpp]: OK I will remove this test. I have refactored some things with the latest update, and included a test for the batching async code independent of the stats calling it, so that should provide good coverage that the batching / async behavior is working as expected and we can avoid a potentially risky test like this one.

File: src/windows/wslc/tasks/ContainerTasks.cpp (hunk 4 above)
Comment (dkbennett) [reply src/windows/wslc/tasks/ContainerTasks.cpp]: Latest iteration does batches of requests in parallel so we can handle the common case very rapidly but protect against extreme cases of very large container counts that might overly stress the endpoints. I put in a batch size of 10 by default but its configurable.

File: src/windows/wslc/tasks/ContainerTasks.cpp [hunk 5]
```diff
@@ -61,6 +62,80 @@ std::string FormatBytes(uint64_t bytes)
...
+        static_cast<double>(stats.cpu_stats.cpu_usage.total_usage) - static_cast<double>(stats.precpu_stats.cpu_usage.total_usage);
+    const auto systemDelta = static_cast<double>(stats.cpu_stats.system_cpu_usage) - static_cast<double>(stats.precpu_stats.system_cpu_usage);
+
+    // When online_cpus is 0 (older API responses), fall back to percpu_usage array length — matches Docker CLI behavior.
+    uint32_t onlineCpus = stats.cpu_stats.online_cpus;
+    if (onlineCpus == 0)
+    {
+        onlineCpus = stats.cpu_stats.cpu_usage.percpu_usage.has_value()
+                         ? static_cast<uint32_t>(stats.cpu_stats.cpu_usage.percpu_usage->size())
+                         : 1u;
```
[the remark this answers is not in the document]
Comment (dkbennett) [reply src/windows/wslc/tasks/ContainerTasks.cpp]: This code was removed so resolving as moot.

File: src/windows/wslc/tasks/ContainerTasks.cpp [hunk 6]
```diff
@@ -495,100 +570,37 @@ void ShowContainerStats(CLIExecutionContext& context)
...
-            stats = ContainerService::Stats(session, WideToMultiByte(containerId));
-        }
-        catch (const wil::ResultException& ex)
-        {
+    wsl::windows::wslc::ForEachAsync<std::wstring>(
+        containers,
+        // Work
+        [&session](const std::wstring& containerId) {
+            return ComputeContainerStatsJson(ContainerService::Stats(session, WideToMultiByte(containerId)));
+        },
```
[the remark this answers is not in the document]
Comment (dkbennett) [reply src/windows/wslc/tasks/ContainerTasks.cpp:580]: Good to be thorough! No change is required here because the WSLCSessionReference implements FtmBase, so the COM objects are free-threaded and agile. Valid concern but doesn't apply here.

I have updated the documentation for ForEachAsync to add a COM threading note about this and indicate why the WSLC interfaces are safe to use it without COM initialization calls.

File: src/windows/wslc/core/AsyncExecution.h [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+    using TResult = decltype(onWork(std::declval<TItem>()));
+
+    struct BatchResult
+    {
+        TItem item;
+        std::optional<TResult> result;
+        wil::ResultException error{S_OK};
+        bool hasError{false};
+    };
```
[the remark this answers is not in the document]
Comment (dkbennett) [reply src/windows/wslc/core/AsyncExecution.h:57]: Good catch, updated to explicitly construct BatchResult with a move.

File: src/windows/wslc/core/AsyncExecution.h [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+// This keeps wall time proportional to ceil(N / batchSize) rather than N for operations
+// that have inherent per-item latency (e.g. network or IPC calls).
+//
+// TWork   : TItem -> TResult                              (called concurrently)
+// TSuccess: TResult -> void                               (called serially)
+// TError  : (TItem, wil::ResultException) -> void         (called serially)
+template <typename TItem, typename TWork, typename TSuccess, typename TError>
+void ForEachAsync(const std::vector<TItem>& items, TWork onWork, TSuccess onSuccess, TError onError, size_t batchSize = 10)
+{
+    WI_ASSERT(batchSize > 0);
```
[the remark this answers is not in the document]
Comment (dkbennett) [reply src/windows/wslc/core/AsyncExecution.h:42]: Not necessary for the current usage, but doesn't hurt for sanity and in case this code is used later for other purposes which might have variable batch sizes, so I added this suggestion.

File: src/windows/wslc/core/AsyncExecution.h [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+Abstract:
+
+    Provides ForEachAsync, a generic helper for executing a work callback
+    over a collection concurrently in bounded batches using std::async.
+
+--*/
+#pragma once
+
+#include <future>
+#include <optional>
```
[the remark this answers is not in the document]
Comment (dkbennett) [reply src/windows/wslc/core/AsyncExecution.h:19]: Added.

File: src/windows/wslc/tasks/ContainerTasks.cpp (hunk 2 above)
Comment (dkbennett) [reply src/windows/wslc/tasks/ContainerTasks.cpp]: It doesn't happen with our version, so I removed. I initially had it in becuase we downgraded the version recently so it was defense against a future change where we might downgrade again but that seems extremely unlikely so I removed it.

Comment (dkbennett) [reply src/windows/wslc/tasks/ContainerTasks.cpp]: Welp scratch that, Copilot complained that it doesn't match docker's implementation exactly if I don't have that, so I added a slight variation back in so it matches docker for correctness sake.

File: src/windows/wslc/core/AsyncExecution.h [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+// This keeps wall time proportional to ceil(N / batchSize) rather than N for operations
+// that have inherent per-item latency (e.g. network or IPC calls).
+//
+// COM threading note: worker threads created by std::async are not guaranteed to have
+// COM initialized. Callers whose onWork callback uses COM interfaces must ensure those
+// interfaces are free-threaded (agile / FtmBase) and that the process has an active MTA
+// (e.g. via CoIncrementMTAUsage or RO_INIT_MULTITHREADED) so that cross-thread calls are
+// valid without per-thread CoInitializeEx. The WSLC session interfaces (IWSLCSession,
+// IWSLCContainer) implement FtmBase and the wslc process initializes the MTA, satisfying
+// this requirement.
```
[the remark this answers is not in the document]
Comment (dkbennett) [reply src/windows/wslc/core/AsyncExecution.h]: Two copilot agents give conflicting information - which one was correct? This one! Updated comment and added com initialization to the worker so we don't initialize COM in cases where it is not required.

File: src/windows/wslc/tasks/ContainerTasks.cpp [hunk 11]
```diff
@@ -495,100 +560,37 @@ void ShowContainerStats(CLIExecutionContext& context)
...
```
[the remark this answers is not in the document]
Comment (dkbennett) [reply src/windows/wslc/tasks/ContainerTasks.cpp:580]: Fixed with wil wrapper for initialization and cleanup consistent with other uses in this codebase.

File: src/windows/wslc/tasks/ContainerTasks.cpp [hunk 12]
```diff
@@ -61,6 +62,70 @@ std::string FormatBytes(uint64_t bytes)
...
+{
+    // Calculate CPU %
+    // Formula matches Docker CLI: https://github.com/docker/cli/blob/master/cli/command/container/stats_helpers.go
+    double cpuPercent = 0.0;
+    const auto cpuDelta =
+        static_cast<double>(stats.cpu_stats.cpu_usage.total_usage) - static_cast<double>(stats.precpu_stats.cpu_usage.total_usage);
+    const auto systemDelta = static_cast<double>(stats.cpu_stats.system_cpu_usage) - static_cast<double>(stats.precpu_stats.system_cpu_usage);
+    if (systemDelta > 0.0 && cpuDelta > 0.0)
+    {
+        cpuPercent = (cpuDelta / systemDelta) * static_cast<double>(stats.cpu_stats.online_cpus) * 100.0;
```
[the remark this answers is not in the document]
Comment (dkbennett) [reply src/windows/wslc/tasks/ContainerTasks.cpp]: Fine I'll put it back, it should now match docker algorithm exactly. Until you run again and find something else wrong with it.

File: src/windows/wslc/core/AsyncExecution.h [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+                    result.hasError = true;
+                    result.error = ex;
+                }
+                return result;
+            }));
+        }
+
+        for (auto& future : futures)
+        {
+            auto batchResult = future.get();
```
Comment (OneBlue) [inline src/windows/wslc/core/AsyncExecution.h:86]: If I'm reading this correctly, this will block until all items from a batch are completed (so if 9 / 10 items are completed, we'll still wait on the 10th items to complete before launching new tasks). 

This isn't a blocker for this PR, but if we make this a generic helper, we should probably fix this. The easiest way I can think of would be to create `batchSize` threads and use `WaitForMultipleObject()` to pick up the next completed thread and give it a new task once completed

**Review by OneBlue — APPROVED** [review]

File: src/windows/wslc/core/AsyncExecution.h (hunk 13 above)
Comment (dkbennett) [reply src/windows/wslc/core/AsyncExecution.h:86]: I'll work on a specific fix for this separately, also add a timeout so we don't wait forever.
